### PR TITLE
file transfer method removal capability

### PR DESCRIPTION
**Pull request by josephsawaya**
This commit adds the file removal feature
to the file transfer method.

Fixes: #61 

Signed-off-by: Joseph Sawaya <jsawaya@redhat.com>

### CODE REVIEW

**Review by cooktheryan — APPROVED** [review]

File: pkg/engine/filetransfer.go [hunk 1]
```diff
@@ -3,49 +3,80 @@ package engine
...
-	file := filepath.Base(path)
+const stopped = define.ContainerStateStopped
 
-	source := filepath.Join("/opt", path)
-	copyFile := (source + " " + dest)
+func fileTransferPodman(ctx context.Context, path, dest, method string, target *api.Target, prev *string) error {
 
 	// Create a new Podman client
 	conn, err := bindings.NewConnection(ctx, "unix://run/podman/podman.sock")
 	if err != nil {
```
Comment (jeffvance) [inline pkg/engine/filetransfer.go:26]: question: `conn` above is a connection, not a context, right?

File: pkg/engine/filetransfer.go [hunk 2]
```diff
@@ -56,26 +87,68 @@ func fileTransferPodman(ctx context.Context, path, dest, method string, target *
...
+		NSMode: "host",
+		Value:  "",
+	}
+	s.Command = []string{"sh", "-c", "rm " + pathToRemove}
+	s.Mounts = []specs.Mount{{Source: dest, Destination: dest, Type: "bind", Options: []string{"rw"}}}
+	s.Volumes = []*specgen.NamedVolume{{Name: harpoonVolume, Dest: "/opt", Options: []string{"ro"}}}
+	return s
+}
+
+func fetchImage(conn context.Context) error {
```
Comment (jeffvance) [inline pkg/engine/filetransfer.go:107]: I think it's better to name all go context variables as `ctx`. `conn` is easily confused to be a connection. k8s uses `ctx` almost a lot!

File: pkg/engine/filetransfer.go [hunk 3]
```diff
@@ -56,26 +87,68 @@ func fileTransferPodman(ctx context.Context, path, dest, method string, target *
...
+	present, err := images.Exists(conn, harpoonImage, nil)
+	klog.Infof("Is image present? %t", present)
 	if err != nil {
 		return err
 	}
+
+	if !present {
```
Comment (jeffvance) [inline pkg/engine/filetransfer.go:114]: nit refactor:
```
if !present && !images.Pull(conn, harpoonImage, nil) {
     return err
}
```

File: pkg/engine/filetransfer.go [hunk 4]
```diff
@@ -56,26 +87,68 @@ func fileTransferPodman(ctx context.Context, path, dest, method string, target *
...
+		_, err = images.Pull(conn, harpoonImage, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func createAndStartContainer(conn context.Context, s *specgen.SpecGenerator) (entities.ContainerCreateResponse, error) {
```
Comment (jeffvance) [inline pkg/engine/filetransfer.go:124]: ditto `ctx` comment...

File: pkg/engine/filetransfer.go [hunk 5]
```diff
@@ -56,26 +87,68 @@ func fileTransferPodman(ctx context.Context, path, dest, method string, target *
...
-	return err
+
+	_, err = containers.Remove(conn, ID, new(containers.RemoveOptions).WithForce(true))
+	if err != nil {
+		// There's a podman bug somewhere that's causing this
+		if err.Error() == "unexpected end of JSON input" {
+			return nil
+		}
+		return err
+	}
```
Comment (jeffvance) [inline pkg/engine/filetransfer.go:151]: another nit refactor:
```
_, err = containers.Remove(conn, ID, new(containers.RemoveOptions).WithForce(true))
if err.Error() != "unexpected end of JSON input" {
      return err
}
return nil
```

**Pull request merged**